service/src/main/java/com/epam/esm/services/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,197 @@
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private CertificateDAOImpl dao;
+    private EntityValidator<Certificate> validator;
+    private CertificateFinder finder;
+    private TagServiceImpl tagServiceImpl;
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDAOImpl dao,
+                                  EntityValidator<Certificate> validator,
+                                  CertificateFinder finder, TagServiceImpl tagServiceImpl) {
+        this.dao = dao;
+        this.validator = validator;
+        this.finder = finder;
+        this.tagServiceImpl = tagServiceImpl;
+    }
+
+    @Override
+    public Certificate create(Certificate certificate) throws ServiceException, ValidationException {
+        try {
+            validator.validate(certificate);
+            return dao.create(certificate);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public Certificate read(int id) throws ServiceException {
+        try {
+            return dao.read(id);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public void delete(int id) throws ServiceException {
+        try {
+            if (dao.read(id) == null) {
+                throw new ServiceException("Entity does not exist!");
+            }
+            dao.delete(id);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public void update(Certificate certificate) throws ServiceException, ValidationException {
+        try {
+            validator.validate(certificate);
+            dao.update(certificate);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> findAll() throws ServiceException {
+        try {
+            return dao.findAll();
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> findBy(EntityFinder<Certificate> entityFinder) throws ServiceException {
+        try {
+            return dao.findBy(entityFinder);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> find(Map<String, String> params) throws ServiceException {
+        finder.newFinder();
+        for (String key : params.keySet()) {
+            try {
+                if (key.contains("sort")) {
+                    finder.sortBy(CertificateSortingParameters.getEntryByParameter(key).getValue(),
+                            parseAscDesc(params.get(key)));
+                } else {
+
+                    CertificateSearchParameters parameter = CertificateSearchParameters.getEntryByParameter(key);
+                    switch (parameter) {
```

Comment:
enstead switch , in method find can be added parameter with columnname and use the unify approach.

in find method ckeck that params key not empty and then proceed query.

Reply:
`if (key.contains("sort")) {
                    finder.sortBy(CertificateSortingParameters.getEntryByParameter(key).getValue(),
                            parseAscDesc(params.get(key)));
                } else if (StringUtils.isNotEmpty(key)) {`

>enstead switch , in method find can be added parameter with columnname and use the unify approach.
Variants like PRICE_LESS would not allow such approach. And i thought that it will be to strong connection to persistence layer?